packages/caravan-health/src/wallet.ts

```diff
@@ -0,0 +1,131 @@
+      utxoMassFactor = 0.25;
+    } else if (utxoSetLength >= 15 && utxoSetLength <= 24) {
+      utxoMassFactor = 0.5;
+    } else if (utxoSetLength >= 5 && utxoSetLength <= 14) {
+      utxoMassFactor = 0.75;
+    } else {
+      utxoMassFactor = 1;
+    }
+    return utxoMassFactor;
+  }
+
+  /*
+    Utility function that helps to obtain the fee rate of the transaction
+  */
+  getFeeRateForTransaction(transaction: Transaction): number {
+    let fees: number = transaction.fee;
+    let weight: number = transaction.weight;
+    return fees / weight;
+  }
+
+  /*
+    Utility function that helps to obtain the percentile of the fees paid by user in tx block
+  */
+  getFeeRatePercentileScore(
+    timestamp: number,
+    feeRate: number,
+    feeRatePercentileHistory: FeeRatePercentile[],
+  ): number {
+    let percentile: number = this.getClosestPercentile(
+      timestamp,
+      feeRate,
+      feeRatePercentileHistory,
+    );
+    return 1 - percentile / 100;
+  }
+
+  /*
+    Utility function that helps to obtain the closest percentile of the fees paid by user in tx block
+  */
+  getClosestPercentile(
+    timestamp: number,
+    feeRate: number,
+    feeRatePercentileHistory: FeeRatePercentile[],
+  ): number {
+    // Find the closest entry by timestamp
+    let closestBlock: FeeRatePercentile | null = null;
+    let closestDifference: number = Infinity;
+
+    for (const block of feeRatePercentileHistory) {
+      const difference = Math.abs(block.timestamp - timestamp);
+      if (difference <= closestDifference) {
+        closestDifference = difference;
+        closestBlock = block;
+      }
+    }
+    if (!closestBlock) {
+      throw new Error("No fee rate data found");
+    }
+    // Find the closest fee rate percentile
+    switch (true) {
+      case feeRate <= closestBlock.avgFee_0:
+        return 0;
+      case feeRate <= closestBlock.avgFee_10:
+        return 10;
+      case feeRate <= closestBlock.avgFee_25:
+        return 25;
+      case feeRate <= closestBlock.avgFee_50:
+        return 50;
+      case feeRate <= closestBlock.avgFee_75:
+        return 75;
+      case feeRate <= closestBlock.avgFee_90:
+        return 90;
+      case feeRate <= closestBlock.avgFee_100:
+        return 100;
+      default:
+        throw new Error("Invalid fee rate");
+    }
+  }
+
+  constructAddressUsageMap(transactions: Transaction[]): AddressUsageMap {
+    let addressUsageMap: AddressUsageMap = {};
+    for (const tx of transactions) {
+      for (const output of tx.vout) {
+        let address = output.scriptPubkeyAddress;
+        if (addressUsageMap[address]) {
+          addressUsageMap[address] = true;
+        } else {
+          addressUsageMap[address] = false;
+        }
+      }
+    }
+    return addressUsageMap;
+  }
+
+  /* 
+    Utility function to check if the given address was used already in past transactions
+  */
+  isReusedAddress(address: string, addressUsageMap: AddressUsageMap): boolean {
```

Comment:
This would be a cleaner API imo and would leverage the class pattern better to just take an address as an arg but take the usage map from a property on the class and then do `if (this.addressUsageMap.has(address)) return true`